WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/ShippingLabelsTopBannerFactory.swift

```diff
@@ -0,0 +1,113 @@
+import Yosemite
+
+/// Creates a top banner for shipping labels asynchronously based on the shipping labels and app settings.
+/// The top banner has two actions: "give feedback" and "dismiss".
+/// This class is not meant to be retained since it uses strong `self` in action handling, so that it has the same life cycle as the top banner.
+final class ShippingLabelsTopBannerFactory {
+    private let shippingLabels: [ShippingLabel]
+    private let stores: StoresManager
+    private let analytics: Analytics
+
+    init(shippingLabels: [ShippingLabel],
+         stores: StoresManager = ServiceLocator.stores,
+         analytics: Analytics = ServiceLocator.analytics) {
+        self.shippingLabels = shippingLabels
+        self.stores = stores
+        self.analytics = analytics
+    }
+
+    /// Creates a top banner asynchronously based on the shipping labels and app settings.
+    /// - Parameters:
+    ///   - isExpanded: whether the top banner is expanded by default.
+    ///   - expandedStateChangeHandler: called when the top banner expanded state changes.
+    ///   - onGiveFeedbackButtonPressed: called when the give feedback button is pressed.
+    ///   - onDismissButtonPressed: called when the dismiss button is pressed.
+    ///   - onCompletion: called when it finishes determining whether the top banner is being shown, and an optional top banner is returned.
+    func createTopBannerIfNeeded(isExpanded: Bool,
+                                 expandedStateChangeHandler: @escaping () -> Void,
+                                 onGiveFeedbackButtonPressed: @escaping () -> Void,
+                                 onDismissButtonPressed: @escaping () -> Void,
+                                 onCompletion: @escaping (TopBannerView?) -> Void) {
+        determineIfTopBannerShouldBeShown { [weak self] shouldShow in
+            guard let self = self else { return }
+
+            guard shouldShow else {
+                onCompletion(nil)
+                return
+            }
+
+            onCompletion(self.createTopBanner(isExpanded: isExpanded,
+                                              expandedStateChangeHandler: expandedStateChangeHandler,
+                                              onGiveFeedbackButtonPressed: onGiveFeedbackButtonPressed,
+                                              onDismissButtonPressed: onDismissButtonPressed))
+        }
+    }
+}
+
+private extension ShippingLabelsTopBannerFactory {
+    func determineIfTopBannerShouldBeShown(onCompletion: @escaping (_ shouldShow: Bool) -> Void) {
+        guard shippingLabels.nonRefunded.isNotEmpty else {
+            onCompletion(false)
+            return
+        }
+
+        let action = AppSettingsAction.loadFeedbackVisibility(type: .shippingLabelsRelease1) { result in
```

Comment:
Not related to this PR, but having this kind of actions synchronous would make our lives much easier, don't you think? 😅

Reply:
Yes I agree 😓 if the operation has to be async, it'd be nice to have something like `async/await` like in Javascript.

after some quick search, it might actually be available in Swift in the future 😮 https://github.com/apple/swift-evolution/blob/main/proposals/0296-async-await.md